### PR TITLE
Feature/gazebo lidar plugins

### PR DESCRIPTION
**Pull request by zarmomin**
Adds parametrized and extendable URDF model of a Velodyne HDL-32 Lidar. Further implements a control plugin to spin the Lidar at constant speed as well as a plugin to publish the beams in XYZ coordinates while maintaining angular information.
Adds a launch file to demonstrate it: `roslaunch rotors_description xacrodisplay_velodyne.launch`
Optionally show the assembled pointcloud: `roslaunch rotors_description xacrodisplay_velodyne.launch assembler_enabled:=True`

### CODE REVIEW

**Review by fmina — CHANGES_REQUESTED** [review]
Minor changes. 
It doesn't pass all the checks to merge. Try to merge locally first to find conflicts.

**Comment by zarmomin** [comment]
Regarding the checks: It appears that Jenkins fails on 14.04 since it cannot locate libgeographic-dev, and the last successful build was > 2 months ago.

File: rotors_description/rviz/hdl32.rviz [hunk 1]
```diff
@@ -0,0 +1,196 @@
+Panels:
```
[the remark this answers is not in the document]
Comment (zarmomin) [reply rotors_description/rviz/hdl32.rviz]: So into rotors_description/resources/rviz/...?

File: rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  // having one joint that is the rotational joint.
+  if (_sdf->HasElement("joint")) {
+    std::string jointName = _sdf->Get<std::string>("joint");
+    this->joint_ = _model->GetJoint(jointName);
+  } else {
+    this->joint_ = _model->GetJoints()[0];
+    ROS_INFO("No joint name provided, therefore the first joint is used.");
+  }
+  // Setup a P-controller, with a gain of 0.1.
+  this->pid_ = common::PID(0.1, 0, 0);
```
[the remark this answers is not in the document]
Comment (zarmomin) [reply rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp]: I'll make all of them adjustable, thanks for pointing that out!

**Comment by zarmomin** [comment]
@fmina @inkyusa I added a couple of things, it'd be great if you could take another look.

File: rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                      current_motor_state_.velocity_limit_rad_s, 100.0);
+  getSdfParam<double>(sdf, "default_torque_limit",
+                      current_motor_state_.torque_limit, 10);
+  getSdfParam<std::string>(sdf, "frequency_service_name", service_name_,
+                           "set_frequency");
+
+  std::string joint_state_topic_name;
+  getSdfParam<std::string>(sdf, "joint_state_topic_name",
+                           joint_state_topic_name, "joint_state");
+  joint_->SetPosition(0, current_motor_state_.current_pos_rad);
```
Comment (fmina) [inline rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp]: SetPosition, GetParam, SetParam seem to be incompatible with gazebo 2, that's why Jenkins fail on ubuntu trusty.
`trusty/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp: In member function 'virtual void gazebo::ContinousJointControllerPlugin::Load(gazebo::physics::ModelPtr, sdf::ElementPtr)':
/home/jenkins/workspace/rotors_simulator/osx/ubuntu-trusty/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp:92:11: error: 'class gazebo::physics::Joint' has no member named 'SetPosition'
   joint_->SetPosition(0, current_motor_state_.current_pos_rad);
           ^
/home/jenkins/workspace/rotors_simulator/osx/ubuntu-trusty/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp: In member function 'void gazebo::ContinousJointControllerPlugin::UpdateMotor(const MotorState&)':
/home/jenkins/workspace/rotors_simulator/osx/ubuntu-trusty/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp:139:15: error: 'class gazebo::physics::Joint' has no member named 'GetParam'
   if (joint_->GetParam("vel", 0) != current_motor_state_.velocity_rad_s) {
               ^
/home/jenkins/workspace/rotors_simulator/osx/ubuntu-trusty/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp:140:13: error: 'class gazebo::physics::Joint' has no member named 'SetParam'
     joint_->SetParam("vel", 0, current_motor_state_.velocity_rad_s);
             ^
/home/jenkins/workspace/rotors_simulator/osx/ubuntu-trusty/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp:145:13: error: 'class gazebo::physics::Joint' has no member named 'SetParam'
     joint_->SetParam("fmax", 0, read_motor_state.torque_limit);
             ^
/home/jenkins/workspace/rotors_simulator/osx/ubuntu-trusty/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp:147:13: error: 'class gazebo::physics::Joint' has no member named 'SetParam'
     joint_->SetParam("fmax", 0, 0.0);`

**Comment by fmina** [comment]
minor comment on Jenkins failure, otherwise LGTM

File: rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp (hunk 3 above)
Comment (zarmomin) [reply rotors_gazebo_plugins/src/gazebo_ros_continous_joint_controller_plugin.cpp]: Thanks, it got farther than it did last time. I updated the CMakeLists to skip this plugin if the gazebo version is too low

File: rotors_description/urdf/velodyne_VLP_16.urdf.xacro [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    </joint>
+
+    <link name="${namespace}/${lidar_base_frame}">
+      <inertial>
+        <mass value="${m1}"/>
+        <origin rpy="0 0 0" xyz="0 0 ${l1*0.5}"/>
+        <inertia ixx="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" ixy="0" ixz="0"
+                 iyy="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" iyz="0"
+                 izz="${0.5 * m1 * (r1*r1)}"/>
+      </inertial>
```
Comment (ffurrer) [inline rotors_description/urdf/velodyne_VLP_16.urdf.xacro]: yup same as above

File: rotors_description/urdf/velodyne_VLP_16.urdf.xacro [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    </joint>
+
+    <link name="${namespace}/${lidar_points_frame}">
+      <inertial>
+        <mass value="${m2}"/>
+        <origin rpy="0 0 0" xyz="0 0 0"/>
+        <inertia ixx="${(m2 * (3.0*r2*r2 + l2*l2)) / 12.0}" ixy="0" ixz="0"
+                 iyy="${(m2 * (3.0*r2*r2 + l2*l2)) / 12.0}" iyz="0"
+                 izz="${0.5 * m2 * (r2*r2)}"/>
+      </inertial>
```
Comment (ffurrer) [inline rotors_description/urdf/velodyne_VLP_16.urdf.xacro]: dito

File: rotors_description/urdf/ouster_os1.urdf.xacro [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    </joint>
+
+    <link name="${namespace}/${lidar_points_frame}">
+      <inertial>
+        <mass value="${m2}"/>
+        <origin rpy="0 0 0" xyz="0 0 0"/>
+        <inertia ixx="${(m2 * (3.0*r2*r2 + l2*l2)) / 12.0}" ixy="0" ixz="0"
+                 iyy="${(m2 * (3.0*r2*r2 + l2*l2)) / 12.0}" iyz="0"
+                 izz="${0.5 * m2 * (r2*r2)}"/>
+      </inertial>
```
Comment (ffurrer) [inline rotors_description/urdf/ouster_os1.urdf.xacro]: same

File: rotors_description/urdf/ouster_os1.urdf.xacro [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    </joint>
+
+    <link name="${namespace}/${lidar_base_frame}_hat">
+      <inertial>
+        <mass value="${m1}"/>
+        <origin rpy="0 0 0" xyz="0 0 ${(l1+l2)*0.5}"/>
+        <inertia ixx="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" ixy="0" ixz="0"
+                 iyy="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" iyz="0"
+                 izz="${0.5 * m1 * (r1*r1)}"/>
+      </inertial>
```
Comment (ffurrer) [inline rotors_description/urdf/ouster_os1.urdf.xacro]: same

File: rotors_description/urdf/ouster_os1.urdf.xacro [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    </joint>
+
+    <link name="${namespace}/${lidar_base_frame}">
+      <inertial>
+        <mass value="${m1}"/>
+        <origin rpy="0 0 0" xyz="0 0 ${l1*0.5}"/>
+        <inertia ixx="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" ixy="0" ixz="0"
+                 iyy="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" iyz="0"
+                 izz="${0.5 * m1 * (r1*r1)}"/>
+      </inertial>
```
Comment (ffurrer) [inline rotors_description/urdf/ouster_os1.urdf.xacro]: Maybe create something like https://github.com/ethz-asl/rotors_simulator/blob/master/rotors_description/urdf/multirotor_base.xacro#L31

File: rotors_description/urdf/velodyne_VLP_16.urdf.xacro [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    </joint>
+
+    <link name="${namespace}/${lidar_base_frame}_hat">
+      <inertial>
+        <mass value="${m1}"/>
+        <origin rpy="0 0 0" xyz="0 0 ${(l1+l2)*0.5}"/>
+        <inertia ixx="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" ixy="0" ixz="0"
+                 iyy="${(m1 * (3.0*r1*r1 + l1*l1)) / 12.0}" iyz="0"
+                 izz="${0.5 * m1 * (r1*r1)}"/>
+      </inertial>
```
Comment (ffurrer) [inline rotors_description/urdf/velodyne_VLP_16.urdf.xacro]: same

**Review by ffurrer — APPROVED** [review]
lgtm, please run clang-format before merging

**Review by fmina — APPROVED** [review]
LGTM

**Comment by michaelpantic** [comment]
We are changing our codebase to gazebo9 soon. Please wait with merging this. 
@zarmomin  can you address the comments of @ffurrer until mid next week? Then I would rebase this to gazebo9 branch and test it.

**Comment by zarmomin** [comment]
@michaelpantic I addressed Fadris comments in https://github.com/ethz-asl/rotors_simulator/pull/527/commits/bf67a530b4c31aaad0c44c8b1e758dcdef319887 and https://github.com/ethz-asl/rotors_simulator/pull/527/commits/44910db2c58121b62579e0109e0f6ad18a0b2756
You should be good to go with gazebo 9 :-)

**Comment by marselomeri** [comment]
having problems building this~
**Conflicting files** 
gazebo_os1_plugin.h
gazebo_os1_plugin.cpp
Screenshot from 2019-04-29 19-05-35

**Comment by zarmomin** [comment]
It looks like you have quite a lot of files from other branches in there (all the .bak extensions etc). If you check out this PR/branch and remove all those remaining files in git it should build just fine.

**Comment by michaelpantic** [comment]
@zarmomin Can you merge in current master into this branch? Then we can check this again and merge.

**Pull request closed without merging**